File: windows/WhisperWin/Core/DictionaryStore.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text.Encodings.Web;
using System.Text.Json;

namespace WhisperWin.Core
{
    /// <summary>
    /// Read-modify-write store for shared/dictionary.json — mirrors Sources/DictionaryStore.swift
    /// so both apps apply the same safety rules when the Settings UI edits the file (previously
    /// this file was read-only on Windows; Phase 2 adds the editor that writes it back).
    /// </summary>
    public static class DictionaryStore
    {
        private static readonly JsonSerializerOptions WriteOptions = new()
        {
            WriteIndented = true,
            // shared/dictionary.json is meant to be human-readable, including Thai text — the
            // default encoder escapes non-ASCII as \uXXXX, which the Mac side (withoutEscapingSlashes,
            // prettyPrinted) does not do. UnsafeRelaxedJsonEscaping keeps output identical in spirit.
            Encoder = JavaScriptEncoder.UnsafeRelaxedJsonEscaping,
        };

        /// <summary>
        /// Outcome of attempting to read the on-disk file — distinguishes "nothing there yet"
        /// (first run, safe to start from a blank file) from "something is there but we couldn't
        /// read it" (OneDrive sync lock, a crashed write, corruption). The "unreadable" case must
        /// NOT be treated as empty, or a read-modify-write save would silently overwrite the
        /// user's real data.
        /// </summary>
        private enum ReadOutcome { Ok, Missing, Unreadable }

        private static ReadOutcome TryRead(string path, out DictionaryFile file)
        {
            file = new DictionaryFile();
            if (!File.Exists(path))
            {
                return ReadOutcome.Missing;
            }

            try
            {
                var json = File.ReadAllText(path);
                var parsed = JsonSerializer.Deserialize<DictionaryFile>(json);
                if (parsed == null)
                {
                    return ReadOutcome.Unreadable;
                }
                file = parsed;
                return ReadOutcome.Ok;
            }
            catch (JsonException)
            {
                return ReadOutcome.Unreadable;
            }
            catch (IOException)
            {
                return ReadOutcome.Unreadable;
            }
        }

        /// <summary>
        /// Reads the file for display in the Settings UI. Falls back to an empty DictionaryFile on
        /// any failure — including "unreadable" — so a locked/corrupt file never crashes the
        /// window; it just temporarily shows nothing until the file is readable again. This is a
        /// read-only path, so falling back here can't destroy anything on disk.
        /// </summary>
        public static DictionaryFile Load(string path)
        {
            return TryRead(path, out var file) == ReadOutcome.Ok ? file : new DictionaryFile();
        }

        private static void Write(string path, DictionaryFile file)
        {
            var directory = Path.GetDirectoryName(path);
            if (!string.IsNullOrEmpty(directory))
            {
                Directory.CreateDirectory(directory);
            }

            var json = JsonSerializer.Serialize(file, WriteOptions);
            File.WriteAllText(path, json);
        }

        /// <summary>
        /// Read-modify-write: replaces <c>entries</c> while preserving whatever <c>pairs</c> is on
        /// disk right now, so saving the word list never clobbers the replacement pairs (and vice
        /// versa in <see cref="SavePairs"/>). This also shrinks the race window against the Mac app
        /// writing the same file through OneDrive — we always start from the latest bytes on disk.
        ///
        /// Safety: if the file exists but can't be read/parsed right now (sync lock, partial write,
        /// corruption), this is a no-op — writing back would silently delete the user's real
        /// entries/pairs. Only "file doesn't exist yet" is treated as a blank slate.
        /// </summary>
        public static void SaveEntries(string path, List<string> entries)
        {
            switch (TryRead(path, out var file))
            {
                case ReadOutcome.Unreadable:
                    return;
                case ReadOutcome.Missing:
                    Write(path, new DictionaryFile { Entries = entries, Pairs = new List<DictionaryPair>() });
                    return;
                case ReadOutcome.Ok:
                    file.Entries = entries;
                    Write(path, file);
                    return;
            }
        }

        public static void SavePairs(string path, List<DictionaryPair> pairs)
        {
            switch (TryRead(path, out var file))
            {
                case ReadOutcome.Unreadable:
                    return;
                case ReadOutcome.Missing:
                    Write(path, new DictionaryFile { Entries = new List<string>(), Pairs = pairs });
                    return;
                case ReadOutcome.Ok:
                    file.Pairs = pairs;
                    Write(path, file);
                    return;
            }
        }
    }
}

File: Sources/DictionaryStore.swift
import Foundation

/// A user-defined "wrong word → right word" replacement pair — for STT mishears that can't be
/// guessed from context (e.g. "ติมสแกน" → "Theme Scan"). Shared with the Windows sibling app via
/// shared/dictionary.json ({ "to_replace", "replace_with" }).
struct Pair: Codable, Equatable, Hashable {
    var toReplace: String
    var replaceWith: String

    enum CodingKeys: String, CodingKey {
        case toReplace = "to_replace"
        case replaceWith = "replace_with"
    }
}

/// Custom dictionary — names/terms that must always be spelled exactly as written.
/// Stored in <repo>/shared/dictionary.json so the Windows sibling app (WhisperWin)
/// uses the same list. Falls back to ~/.whisperapp/dictionary.json when the app
/// runs outside the repo (e.g. installed in /Applications).
enum DictionaryStore {
    static let seedEntries = [
        "Tar Sawang",
        "Coffee for Worker",
        "Claude Code",
        "Wispr Flow",
        "Take Home",
        "Micro Level",
        "Content Direction",
        "tier list",
    ]

    /// `pairs` is optional so files written before this feature (just `entries`) still decode.
    private struct FileFormat: Codable { var entries: [String]; var pairs: [Pair]? }

    /// Walk up from the app bundle looking for the repo's shared/dictionary.json
    private static func sharedPath() -> String? {
        var dir = URL(fileURLWithPath: Bundle.main.bundlePath).deletingLastPathComponent()
        for _ in 0..<5 {
            let candidate = dir.appendingPathComponent("shared/dictionary.json")
            if FileManager.default.fileExists(atPath: candidate.path) { return candidate.path }
            dir.deleteLastPathComponent()
        }
        return nil
    }

    private static var fallbackPath: String { KeyStore.dir + "/dictionary.json" }

    static var activePath: String { sharedPath() ?? fallbackPath }

    /// Outcome of attempting to read the on-disk file — distinguishes "nothing there yet"
    /// (normal first run, safe to seed) from "something is there but we couldn't read it"
    /// (OneDrive sync lock, a crashed write, corruption — must NOT be treated as empty, or a
    /// read-modify-write save would silently overwrite the user's real data with the seed).
    private enum ReadResult {
        case ok(FileFormat)
        case missing
        case unreadable
    }

    private static func attemptRead() -> ReadResult {
        guard FileManager.default.fileExists(atPath: activePath) else { return .missing }
        guard let data = FileManager.default.contents(atPath: activePath) else { return .unreadable }
        guard let file = try? JSONDecoder().decode(FileFormat.self, from: data) else { return .unreadable }
        return .ok(file)
    }

    /// Reads the file for display/use (correction prompt assembly, populating Settings UI).
    /// Falls back to the seed on any failure — including "unreadable" — so a locked/corrupt file
    /// never blocks dictation; it just temporarily loses the custom entries until it's readable
    /// again. This is a read-only path, so falling back here can't destroy anything on disk.
    private static func readFile() -> FileFormat {
        if case .ok(let file) = attemptRead() { return file }
        return FileFormat(entries: seedEntries, pairs: [])
    }

    private static func writeFile(_ file: FileFormat) {
        let enc = JSONEncoder()
        enc.outputFormatting = [.prettyPrinted, .withoutEscapingSlashes]
        guard let data = try? enc.encode(file) else { return }
        if sharedPath() == nil {
            try? FileManager.default.createDirectory(atPath: KeyStore.dir, withIntermediateDirectories: true)
        }
        try? data.write(to: URL(fileURLWithPath: activePath))
    }

    static func load() -> [String] {
        let file = readFile()
        return file.entries.isEmpty ? seedEntries : file.entries
    }

    static func loadPairs() -> [Pair] {
        readFile().pairs ?? []
    }

    /// Read-modify-write: replaces `entries` while preserving whatever `pairs` are on disk, so
    /// saving the word list never clobbers the replacement pairs (and vice versa in savePairs).
    ///
    /// Safety: if the file exists but can't be read/decoded right now (sync lock, partial write,
    /// corruption), this is a no-op — writing the seed back would silently delete the user's real
    /// entries/pairs. Only "file doesn't exist yet" (first run) is treated as a blank slate.
    static func saveEntries(_ entries: [String]) {
        switch attemptRead() {
        case .unreadable:
            return
        case .missing:
            writeFile(FileFormat(entries: entries, pairs: []))
        case .ok(var file):
            file.entries = entries
            writeFile(file)
        }
    }

    static func savePairs(_ pairs: [Pair]) {
        switch attemptRead() {
        case .unreadable:
            return
        case .missing:
            writeFile(FileFormat(entries: seedEntries, pairs: pairs))
        case .ok(var file):
            file.pairs = pairs
            writeFile(file)
        }
    }
}
